samples/matter/lock/src/software_images_swapper.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "software_images_swapper.h"
+
+#include <pm_config.h>
+
+#include <dfu/dfu_multi_image.h>
+#include <dfu/dfu_target.h>
+#include <dfu/dfu_target_mcuboot.h>
+#include <zephyr/sys/reboot.h>
+
+#include <app/server/Server.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_MATTER_LOG_LEVEL);
+
+bool SoftwareImagesSwapper::Swap(ApplicationImageId applicationId)
+{
+	uint32_t app_address;
+	uint32_t app_size;
+	uint32_t net_address;
+	uint32_t net_size;
+
+	if (applicationId == ApplicationImageId::Image_1) {
+		app_address = PM_APP_1_CORE_APP_ADDRESS;
+		app_size = PM_APP_1_CORE_APP_SIZE;
+		net_address = PM_APP_1_CORE_NET_ADDRESS;
+		net_size = PM_APP_1_CORE_NET_SIZE;
+	} else if (applicationId == ApplicationImageId::Image_2) {
+		app_address = PM_APP_2_CORE_APP_ADDRESS;
+		app_size = PM_APP_2_CORE_APP_SIZE;
+		net_address = PM_APP_2_CORE_NET_ADDRESS;
+		net_size = PM_APP_2_CORE_NET_SIZE;
+	} else {
+		LOG_ERR("Requested to swap application image with invalid identifier: %d",
+			static_cast<uint8_t>(applicationId));
+		return false;
+	}
+
+	if (dfu_target_mcuboot_set_buf(mDFUImageBuffer, sizeof(mDFUImageBuffer)) != 0) {
+		return false;
+	}
+
+	if (!SwapImage(app_address, app_size, 0)) {
+		return false;
+	}
+
+	if (!SwapImage(net_address, net_size, 1)) {
+		return false;
+	}
+
+	if (dfu_target_schedule_update(-1) != 0) {
+		return false;
+	}
+
+	/* Wipe out whole settings as they will not apply to the new application image. */
+	chip::Server::GetInstance().ScheduleFactoryReset();
+
+	return true;
+}
+
+bool SoftwareImagesSwapper::SwapImage(uint32_t address, uint32_t size, uint8_t id)
+{
+	LOG_INF("Requested to swap image %d from address %x with size %x", id, address, size);
+
+	if (!mFlashDevice) {
+		LOG_ERR("Invalid flash device");
+		return false;
+	}
+
+	if (dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, id, size, nullptr) != 0) {
+		return false;
+	}
+
+	for (uint32_t offset = 0; offset < size; offset += kBufferSize) {
+		if (flash_read(mFlashDevice, address + offset, mFlashImageBuffer, kBufferSize) != 0) {
```

Comment:
Does the buffer size impact the speed? I wonder if we could reduce the buffer sizes and e.g. move `mFlashImageBuffer` to stack to reduce RAM usage.

Reply:
I think it would be safe to reduce it to e.g. 256 B.

Reply:
Then IMHO it's worth it as we're quite short on RAM recently :)

Reply:
Static buffers were replaced with dynamically allocated buffers of 256 B size each.

Reply:
@kkasperczyk-no I see that the overall swap time is quite long. How it would look like, if we would copy 4096B (1 flash page) not 256? In theory the QSPI memory has up to 48 MB/sec EasyDMA read rate.